File: sim/golden_models/models/RadiogModel.cpp
#include "golden_models/models/RadiogModel.h"

#include "golden_models/core/TestVectorIO.h"

namespace fpd::sim {

namespace {

std::vector<uint16_t> BuildSyntheticDarkFrame(uint32_t frame_index, std::size_t width = 8U) {
    std::vector<uint16_t> frame(width);
    for (std::size_t i = 0; i < width; ++i) {
        frame[i] = static_cast<uint16_t>((frame_index * 11U + static_cast<uint32_t>(i) * 3U) & 0x0FFFU);
    }
    return frame;
}

}  // namespace

void RadiogModel::reset() {
    start_ = 0;
    xray_ready_ = 0;
    xray_on_ = 0;
    xray_off_ = 0;
    dark_frame_mode_ = 0;
    frame_valid_ = 0;
    cfg_dark_cnt_ = 64;
    cfg_tsettle_ = 100;
    cfg_prep_timeout_ = 30;
    state_ = 0;
    xray_enable_ = 0;
    error_ = 0;
    done_ = 0;
    dark_avg_ready_ = 0;
    dark_frames_captured_ = 0;
    timer_ = 0;
    xray_seen_on_ = 0;
    prev_frame_valid_ = 0;
    frame_pixels_.clear();
    dark_accum_.clear();
    avg_dark_frame_.clear();
    cycle_count_ = 0;
}

void RadiogModel::step() {
    done_ = 0;
    dark_avg_ready_ = 0;
    const auto capture_dark_frame = [&]() {
        const std::vector<uint16_t> frame = frame_pixels_.empty()
            ? BuildSyntheticDarkFrame(dark_frames_captured_)
            : frame_pixels_;
        if (dark_accum_.size() != frame.size()) {
            dark_accum_.assign(frame.size(), 0U);
        }
        avg_dark_frame_.resize(frame.size(), 0U);
        for (std::size_t i = 0; i < frame.size(); ++i) {
            dark_accum_[i] += frame[i];
            const uint32_t divisor = dark_frames_captured_ + 1U;
            avg_dark_frame_[i] = static_cast<uint16_t>(dark_accum_[i] / divisor);
        }
        ++dark_frames_captured_;
    };
    switch (state_) {
      case 0:
        if (start_ != 0U) {
            error_ = 0U;
            dark_frames_captured_ = 0U;
            timer_ = 0U;
            xray_seen_on_ = 0U;
            dark_accum_.clear();
            avg_dark_frame_.clear();
            state_ = dark_frame_mode_ != 0U ? 4U : 1U;
            if (dark_frame_mode_ != 0U &&
                (((frame_valid_ != 0U) && (prev_frame_valid_ == 0U)) || frame_pixels_.empty())) {
                capture_dark_frame();
                if (dark_frames_captured_ >= cfg_dark_cnt_) {
                    dark_avg_ready_ = 1U;
                    done_ = 1U;
                    state_ = 0U;
                }
            }
        }
        break;
      case 1:
        if (xray_ready_ != 0U) {
            state_ = 2U;
            xray_enable_ = 1U;
            timer_ = 0U;
        } else if (++timer_ >= cfg_prep_timeout_) {
            error_ = 1U;
            state_ = 7U;
        }
        break;
      case 2:
        xray_enable_ = 1U;
        if (xray_on_ != 0U) {
            xray_seen_on_ = 1U;
        }
        if (xray_seen_on_ != 0U && xray_off_ != 0U) {
            state_ = 3U;
            timer_ = 0U;
        }
        break;
      case 3:
        if (++timer_ >= cfg_tsettle_) {
            done_ = 1U;
            xray_enable_ = 0U;
            state_ = 0U;
        }
        break;
      case 4:
        if (((frame_valid_ != 0U) && (prev_frame_valid_ == 0U)) || frame_pixels_.empty()) {
            capture_dark_frame();
        }
        if (dark_frames_captured_ >= cfg_dark_cnt_) {
            dark_avg_ready_ = 1U;
            done_ = 1U;
            state_ = 0U;
        }
        break;
      default:
        break;
    }
    prev_frame_valid_ = frame_valid_;
    ++cycle_count_;
}

void RadiogModel::set_inputs(const SignalMap& inputs) {
    start_ = GetScalar(inputs, "start", start_);
    xray_ready_ = GetScalar(inputs, "xray_ready", xray_ready_);
    xray_on_ = GetScalar(inputs, "xray_on", xray_on_);
    xray_off_ = GetScalar(inputs, "xray_off", xray_off_);
    dark_frame_mode_ = GetScalar(inputs, "dark_frame_mode", dark_frame_mode_);
    frame_valid_ = GetScalar(inputs, "frame_valid", frame_valid_);
    cfg_dark_cnt_ = GetScalar(inputs, "cfg_dark_cnt", cfg_dark_cnt_);
    cfg_tsettle_ = GetScalar(inputs, "cfg_tsettle", cfg_tsettle_);
    cfg_prep_timeout_ = GetScalar(inputs, "cfg_prep_timeout", cfg_prep_timeout_);
    frame_pixels_ = GetVector(inputs, "frame_pixels");
}

SignalMap RadiogModel::get_outputs() const {
    return {
        {"state", state_},
        {"xray_enable", xray_enable_},
        {"error", error_},
        {"done", done_},
        {"dark_avg_ready", dark_avg_ready_},
        {"dark_frames_captured", dark_frames_captured_},
        {"dark_avg_frame", avg_dark_frame_},
    };
}

std::vector<Mismatch> RadiogModel::compare(const SignalMap& rtl_outputs) const {
    return CompareSignalMaps(cycle(), get_outputs(), rtl_outputs);
}

void RadiogModel::generate_vectors(const std::string& output_dir) {
    TestVectorFile vectors;
    vectors.module_name = "radiography_subfsm";
    vectors.spec_name = "SPEC-FPD-010";
    vectors.clock_name = "sys_clk";
    vectors.signal_inputs = {
        "start", "xray_ready", "xray_on", "xray_off", "dark_frame_mode",
        "frame_valid", "frame_pixels", "cfg_dark_cnt", "cfg_tsettle", "cfg_prep_timeout"
    };
    vectors.signal_outputs = {
        "state", "xray_enable", "error", "done", "dark_avg_ready",
        "dark_frames_captured", "dark_avg_frame"
    };
    reset();
    set_inputs({
        {"start", 1U}, {"xray_ready", 1U}, {"xray_on", 1U},
        {"xray_off", 1U}, {"cfg_tsettle", 2U}
    });
    for (int i = 0; i < 4; ++i) {
        step();
        vectors.vectors.push_back({cycle(), {
            {"start", 1U}, {"xray_ready", 1U}, {"xray_on", 1U}, {"xray_off", 1U}, {"cfg_tsettle", 2U}
        }, get_outputs()});
    }

    reset();
    set_inputs({
        {"start", 1U}, {"dark_frame_mode", 1U}, {"frame_valid", 1U}, {"cfg_dark_cnt", 2U},
        {"frame_pixels", std::vector<uint16_t>{10U, 20U, 30U, 40U}}
    });
    step();
    vectors.vectors.push_back({cycle(), {
        {"start", 1U}, {"dark_frame_mode", 1U}, {"frame_valid", 1U}, {"cfg_dark_cnt", 2U},
        {"frame_pixels", std::vector<uint16_t>{10U, 20U, 30U, 40U}}
    }, get_outputs()});

    set_inputs({
        {"dark_frame_mode", 1U}, {"frame_valid", 1U}, {"cfg_dark_cnt", 2U},
        {"frame_pixels", std::vector<uint16_t>{14U, 24U, 34U, 44U}}
    });
    step();
    vectors.vectors.push_back({cycle(), {
        {"dark_frame_mode", 1U}, {"frame_valid", 1U}, {"cfg_dark_cnt", 2U},
        {"frame_pixels", std::vector<uint16_t>{14U, 24U, 34U, 44U}}
    }, get_outputs()});

    if (GetScalar(get_outputs(), "done") == 0U) {
        step();
        vectors.vectors.push_back({cycle(), {}, get_outputs()});
    }
    WriteHexVectors(vectors, output_dir + "/radiog_vectors.hex");
    WriteBinaryVectors(vectors, output_dir + "/radiog_vectors.bin");
}

}  // namespace fpd::sim

File: sim/viewer/src/FpdSimViewer/Models/RadiogModel.cs
using FpdSimViewer.Models.Core;

namespace FpdSimViewer.Models;

public sealed class RadiogModel : GoldenModelBase
{
    private uint _start;
    private uint _xrayReady;
    private uint _xrayOn;
    private uint _xrayOff;
    private uint _darkFrameMode;
    private uint _frameValid;
    private uint _cfgDarkCnt = 64U;
    private uint _cfgTSettle = 100U;
    private uint _cfgPrepTimeout = 30U;
    private uint _state;
    private uint _xrayEnable;
    private uint _error;
    private uint _done;
    private uint _darkAvgReady;
    private uint _darkFramesCaptured;
    private uint _timer;
    private uint _xraySeenOn;
    private uint _prevFrameValid;
    private ushort[] _framePixels = [];
    private uint[] _darkAccum = [];
    private ushort[] _avgDarkFrame = [];

    public override void Reset()
    {
        _start = 0U;
        _xrayReady = 0U;
        _xrayOn = 0U;
        _xrayOff = 0U;
        _darkFrameMode = 0U;
        _frameValid = 0U;
        _cfgDarkCnt = 64U;
        _cfgTSettle = 100U;
        _cfgPrepTimeout = 30U;
        _state = 0U;
        _xrayEnable = 0U;
        _error = 0U;
        _done = 0U;
        _darkAvgReady = 0U;
        _darkFramesCaptured = 0U;
        _timer = 0U;
        _xraySeenOn = 0U;
        _prevFrameValid = 0U;
        _framePixels = [];
        _darkAccum = [];
        _avgDarkFrame = [];
        CycleCount = 0;
    }

    public override void Step()
    {
        _done = 0U;
        _darkAvgReady = 0U;

        switch (_state)
        {
            case 0U:
                if (_start != 0U)
                {
                    _error = 0U;
                    _darkFramesCaptured = 0U;
                    _timer = 0U;
                    _xraySeenOn = 0U;
                    _darkAccum = [];
                    _avgDarkFrame = [];
                    _state = _darkFrameMode != 0U ? 4U : 1U;
                    if (_darkFrameMode != 0U &&
                        (((_frameValid != 0U) && (_prevFrameValid == 0U)) || _framePixels.Length == 0))
                    {
                        CaptureDarkFrame();
                        if (_darkFramesCaptured >= _cfgDarkCnt)
                        {
                            _darkAvgReady = 1U;
                            _done = 1U;
                            _state = 0U;
                        }
                    }
                }

                break;
            case 1U:
                if (_xrayReady != 0U)
                {
                    _state = 2U;
                    _xrayEnable = 1U;
                    _timer = 0U;
                }
                else if (++_timer >= _cfgPrepTimeout)
                {
                    _error = 1U;
                    _state = 7U;
                }

                break;
            case 2U:
                _xrayEnable = 1U;
                if (_xrayOn != 0U)
                {
                    _xraySeenOn = 1U;
                }

                if (_xraySeenOn != 0U && _xrayOff != 0U)
                {
                    _state = 3U;
                    _timer = 0U;
                }

                break;
            case 3U:
                if (++_timer >= _cfgTSettle)
                {
                    _done = 1U;
                    _xrayEnable = 0U;
                    _state = 0U;
                }

                break;
            case 4U:
                if (((_frameValid != 0U) && (_prevFrameValid == 0U)) || _framePixels.Length == 0)
                {
                    CaptureDarkFrame();
                }

                if (_darkFramesCaptured >= _cfgDarkCnt)
                {
                    _darkAvgReady = 1U;
                    _done = 1U;
                    _state = 0U;
                }

                break;
            default:
                break;
        }

        _prevFrameValid = _frameValid;
        CycleCount++;
    }

    public override void SetInputs(SignalMap inputs)
    {
        _start = SignalHelpers.GetScalar(inputs, "start", _start);
        _xrayReady = SignalHelpers.GetScalar(inputs, "xray_ready", _xrayReady);
        _xrayOn = SignalHelpers.GetScalar(inputs, "xray_on", _xrayOn);
        _xrayOff = SignalHelpers.GetScalar(inputs, "xray_off", _xrayOff);
        _darkFrameMode = SignalHelpers.GetScalar(inputs, "dark_frame_mode", _darkFrameMode);
        _frameValid = SignalHelpers.GetScalar(inputs, "frame_valid", _frameValid);
        _cfgDarkCnt = SignalHelpers.GetScalar(inputs, "cfg_dark_cnt", _cfgDarkCnt);
        _cfgTSettle = SignalHelpers.GetScalar(inputs, "cfg_tsettle", _cfgTSettle);
        _cfgPrepTimeout = SignalHelpers.GetScalar(inputs, "cfg_prep_timeout", _cfgPrepTimeout);
        _framePixels = SignalHelpers.GetVector(inputs, "frame_pixels");
    }

    public override SignalMap GetOutputs()
    {
        return new SignalMap
        {
            ["state"] = _state,
            ["xray_enable"] = _xrayEnable,
            ["frame_valid"] = _frameValid,
            ["error"] = _error,
            ["done"] = _done,
            ["dark_avg_ready"] = _darkAvgReady,
            ["dark_frames_captured"] = _darkFramesCaptured,
            ["frame_pixels"] = _framePixels,
            ["dark_avg_frame"] = _avgDarkFrame,
        };
    }

    private void CaptureDarkFrame()
    {
        var frame = _framePixels.Length == 0 ? BuildSyntheticDarkFrame(_darkFramesCaptured) : _framePixels;
        if (_darkAccum.Length != frame.Length)
        {
            _darkAccum = new uint[frame.Length];
        }

        if (_avgDarkFrame.Length != frame.Length)
        {
            _avgDarkFrame = new ushort[frame.Length];
        }

        for (var index = 0; index < frame.Length; index++)
        {
            _darkAccum[index] += frame[index];
            var divisor = _darkFramesCaptured + 1U;
            _avgDarkFrame[index] = (ushort)(_darkAccum[index] / divisor);
        }

        _darkFramesCaptured++;
    }

    private static ushort[] BuildSyntheticDarkFrame(uint frameIndex, int width = 8)
    {
        var frame = new ushort[width];
        for (var index = 0; index < width; index++)
        {
            frame[index] = (ushort)((frameIndex * 11U + ((uint)index * 3U)) & 0x0FFFU);
        }

        return frame;
    }
}
